File: Java/Rename.java
import java.util.Scanner;
import java.io.File;
import java.io.IOException;

public class Rename {
    public static void main(String[] args) {
        System.out.println("This assumes there are fewer than 10,000 photos to rename");

        File f;
        Scanner input = new Scanner(System.in).useDelimiter("\\n");

        System.out.print("What is the first photo number? ");
        int first = input.nextInt();
        System.out.print("What is the last photo number? ");
        int last = input.nextInt();
	      System.out.println("What number do you want the renamed photos to start at?");
	      int newNumber = input.nextInt();
        String original_prefix = System.console().readLine("What is the current prefix? (DSC)\n");
        String new_prefix = System.console().readLine("What should the prefix be changed to?\n");

        for (int i = 1; i <= last - first + 1; i++) {
            f = new File(fileName(original_prefix, i + first - 1));
            f.renameTo(new File(fileName(new_prefix, i + newNumber - 1)));
        }
    }

    private static String fileName(String prefix, int i) {
        StringBuilder name = new StringBuilder();
        name.append(prefix);
        name.append('_');
        if (i < 10) {
            name.append("000");
        } else if (i < 100) {
            name.append("00");
        } else if (i < 1000) {
            name.append("0");
        }
        name.append(i);
        name.append(".JPG");
        return name.toString();
    }
}

File: rust/src/rename.rs
use rand::distr::{Alphanumeric, SampleString};
use std::env;
use std::fs;

use crate::images::{_get_file_extension, _get_image_extensions, _get_images};
use crate::input::_read_input;

const PREFIX_LENGTH: u32 = 6;

fn _get_new_prefix() -> String {
    println!("What should the new prefix be? Type `.` for the current directory.");
    let mut new_prefix = match _read_input() {
        Ok(prefix) => prefix,
        Err(e) => panic!("Error: failed to read new prefix: {}", e),
    };
    if new_prefix == "." {
        let cwd = match env::current_dir() {
            Ok(cwd) => cwd,
            Err(e) => panic!("Error: failed to get current directory: {}", e),
        };
        let file_name = match cwd.file_name() {
            Some(file_name) => file_name,
            None => panic!("Error: failed to get the top folder from the current directory path."),
        };
        new_prefix = match file_name.to_str() {
            Some(file_name) => file_name.to_string(),
            None => panic!("Error: failed to convert top folder to a string."),
        };
    }
    new_prefix
}

fn _rename_image(image_name: &str, new_prefix: &str, index: u32) {
    if index > 9999 {
        panic!("Error: index out of range - greater than 9999.");
    }
    let extension = _get_file_extension(image_name);
    let new_image_name: String;
    if new_prefix == "" {
        new_image_name = format!("{:04}.{}", index, extension);
    } else {
        new_image_name = format!("{}_{:04}.{}", new_prefix, index, extension);
    }
    println!("Renaming {} to {}", image_name, new_image_name);
    match fs::rename(image_name, new_image_name) {
        Ok(_) => (),
        Err(e) => panic!("Error: failed to rename image: {}", e),
    };
}

fn _rename_all_photos(image_files: Vec<String>, new_prefix: &str, starting_index: &str) {
    let mut index: u32 = starting_index.parse().unwrap();
    for image in image_files {
        _rename_image(&image, new_prefix, index);
        index += 1;
    }
}

fn _generate_random_prefix(length: u32) -> String {
    Alphanumeric.sample_string(&mut rand::rng(), length as usize)
}

pub fn rename() {
    println!("Let's rename some stuff!");
    let new_prefix: String = _get_new_prefix();
    println!("What should the photo numbers start from?");
    let starting_index: String = match _read_input() {
        Ok(prefix) => prefix,
        Err(e) => panic!("Error, failed to read photo start number: {}", e),
    };
    println!(
        "Photos will be renamed with prefix: `{}` and starting index: `{}`",
        new_prefix, starting_index,
    );
    let image_extensions = _get_image_extensions();
    let mut image_files: Vec<String> = _get_images(&image_extensions);
    // rename each file to a random prefix. This ensures that if we keep the same prefix and just
    // update the numbers, nothing will be overwritten
    let rand_prefix: String = _generate_random_prefix(PREFIX_LENGTH);
    _rename_all_photos(image_files, &rand_prefix, &starting_index);
    image_files = _get_images(&image_extensions);
    _rename_all_photos(image_files, &new_prefix, &starting_index);
}
